### ISSUE TITLE
extend torch.util.checkpoint to support offload to cpu/nvme

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#158657]
### 🚀 The feature, motivation and pitch

While activation checkpointing helps a lot with saving memory, as we are moving into huge sequence length those checkpointed hidden_states become huge e.g. llama-8b at 500k seqlen -> `hidden_states.shape: [1, 500_000, 4096]` is `1*500_000*4096*32*2/2**30 = 122 GB`.

This feature request is about extending `torch.util.checkpoint` to support checkpointed tensor offloading to cpu (and optionally nvme as not all users have a lot of cpu memory).

There is a rich discussion about this here:
https://pytorch.slack.com/archives/C3PDTEV8E/p1745274102600729
with @janeyx99's work at [activation offloading](https://github.com/pytorch/torchtune/blob/10c31c0abadf51dfa2bf606637cd81e812e3f8c9/torchtune/training/_activation_offloading.py#L24-L374) at torchtune, which is great, but hardcoded to llama - but would be a great source to adapt into the pytorch core - the non-multi-stream version is generic.

The needs are:
- forward checkpoint offloading is easy to overlap with compute by using stream copy
- the backward requires a prefetch to be non-blocking
- when both forward and backward checkpointing is compute overlapping then an offload to nvme becomes an option for those who don't have enough cpu memory - if there is an overlap the slowness of nvme shouldn't matter

The main use would be in frameworks like HF Transformers where passing an additional flag to activation checkpointing feature would automatically enable the offloading.

In my work https://www.arxiv.org/abs/2506.13996 I was able to do 500K seqlen on llama-8 on a single gpu on H100 and 1.2M on H200. I was using a blocking copy offloading as a proof of concept. Having this a built-in feature in pytorch core would go a long way to help with easier long sequence training in the community.

Thank you.

**Comment by soulitzer** [comment]
Thanks for the nice writeup! If I understand correctly, there are two separate requests here:

1. Bring a generic offloading into core that does streaming/prefetching, etc.
2. Allow enabling offloading on the inputs of AC regions via a flag.

I think we should do both!

To comment on (2) more,  I think generally we can think of offloading it as an extension to selective AC (SAC). Jax, for example, has a offload policy.  If my SAC policy function returns CheckpointPolicy.OFFLOAD for some op, then I would choose to offload that Tensor instead of recomputing /saving.

This request asks for something slightly different though, in that I'm not specifying the policy to an output of an operator of a region, but rather an input to the region, so SAC today cannot actually support this, and so I would agree that we do need a new flag that we would pass to torch.utils.checkpoint. 

Similar to the `context_fn` argument we use today to specify SAC, we could have a similar argument to allow one to specify a context that is enabled at the time the inputs of the checkpoint region are being saved for backward. Maybe a `inputs_context_fn`? e.g. so that once a generic and performant version of offloading lands in core as a saved_tensor_hook, then you can pass it as-is `checkpoint(input_context_fn=offload_to_cpu_saved_tensor_hook)`.

**Comment by stas00** [comment]
Thank you for the feedback, Jeffrey.

I don't think I'm attached to how this is done, so what you feel the best approach would be would work for me. The main "user" would be frameworks like HF Transformers that heavily rely on the current checkpointing mechanism. We can also invite HF Transformers maintainers feedback if it'd be helpful. In their case the user would activate a config feature to enable offloading and they can then either insert the context manager or a flag to the `checkpoint` function.

One important thing that you haven't mentioned in your feedback is the need for a prefetch in the backward path to ensure compute and io overlap. In the forward the async streaming should take care of it, except for the very last layer - there could be some blocking there if the last computation (typically `loss`) is very quick, though I doubt this will ever be the case.

One other important aspect is to check that offloading isn't copying a tied / shared between layers tensor, like an attention mask which often is the same across the layers and which can be very large and could lead to cpu OOM when copied n_layer times.

**Comment by soulitzer** [comment]
Good points. I haven't thought too much about the implementation of a generic offload itself, but maybe @janeyx99 knows more, e.g. what assumptions we made in the TorchTune offloading that "hardcodes" it to llama.

> In their case the user would activate a config feature to enable offloading and they can then either insert the context manager or a flag to the checkpoint function.

Hmm given that the user is operating at a config level, I would wonder how much different offloading is expressed via a context manager over multiple ACs or via a flag to AC itself.  

If I'm essentially adding sugar for

```
with context():
    checkpoint(...)
```

to be expressed as
```
checkpoint(input_context_fn=context_fn)
```
Perhaps it is possible for a layer within HF to handle this?

> And perhaps leaving a slot for supporting NVME offload in the future as well

If NVME offload can be expressed via saved tensor hooks, a generic API like `inputs_context_fn ` should allow it!

**Comment by amoaix20** [comment]
I understand that in @janeyx99’s work on [activation offloading ](https://github.com/pytorch/torchtune/blob/10c31c0abadf51dfa2bf606637cd81e812e3f8c9/torchtune/training/_activation_offloading.py#L24-L374)in torchtune, using OffloadActivations together with torch.utils.checkpoint.checkpoint cannot, within the same layer, simultaneously avoid the recomputation overhead and offload activations to the CPU to save GPU memory. eg:

```python
with OffloadActivations(use_streams=True):
    output = torch.utils.checkpoint.checkpoint(fn, *args, use_reentrant=False)
``` 


@stas00 @soulitzer

**Comment by amoaix20** [comment]
Wouldn't it be easier to understand — and to achieve both avoiding recomputation overhead and offloading activations to the CPU to save GPU memory — by using the context_fn parameter in torch.utils.checkpoint.checkpoint to customize the Caching and Cached operations?
like this :https://github.com/pytorch/pytorch/issues/70135#issuecomment-1542439983

**Comment by amoaix20** [comment]
Use multiple CUDA streams in Caching and Cached class to coordinate overlapping H2D/D2H transfers and computation, enabling asynchronous execution.
@soulitzer @stas00

**Comment by amoaix20** [comment]
Sorry, I didn't make myself clear earlier
First, I agree with your point that from a **user’s perspective**, it’s best not to expose too many technical details — ultimately, controlling the offload behavior through an on/off knob is  user-friendly approach.

However, from a **developer’s perspective**, what confuses me are some of the **implementation details**. Specifically, my understanding of the *checkpoint supporting CPU offload* feature (i.e., **CPU_OFFLOAD**) is that its final behavior should be as follows:

For each transformer block wrapped by `checkpoint`, during the **forward pass**, activations are **offloaded to the CPU (D2H)**, and during the **backward pass**, they are **prefetched back to the GPU (H2D)**. Both data transfers are executed on **a new CUDA stream** (not defult stream) to achieve overlap between computation and data movement.

From the user’s point of view, this design provides two main benefits:

1. **No extra computation time**, because there is no additional forward recomputation during backward (i.e., the latency of `CPU_OFFLOAD` is better than that of `CheckpointPolicy.PREFER_RECOMPUTE`).
2. **No increase in GPU memory footprint**, because the output of the current transformer block is actually stored on the CPU (i.e., the memory footprint of `CPU_OFFLOAD` is smaller than that of `CheckpointPolicy.MUST_SAVE`).

However, Jane’s implementation is based on **`saved_tensors_hooks`**, and when used together with `torch.utils.checkpoint.checkpoint`, it seems to have some issues. For example, in the following code, recomputation still occurs:

```python
with OffloadActivations(use_streams=True):
    output = torch.utils.checkpoint.checkpoint(fn, *args, use_reentrant=False)
```

It seems that avoiding recomputation only works when using `OffloadActivations` **alone**, without `torch.utils.checkpoint.checkpoint`, such as:

```python
with OffloadActivations(use_streams=True):
    output = model(...)
```

Therefore, my idea is to implement new **`CachingMode`** and **`CachedMode`** classes based on **`TorchDispatchMode`**, so that `torch.utils.checkpoint.checkpoint` can introduce a new **`CheckpointPolicy.CPU_OFFLOAD`**.
I believe that building this functionality on **TorchDispatchMode** （[more detail](https://github.com/pytorch/pytorch/issues/70135#issuecomment-1542399767)）is a more robust and reasonable approach than relying on **`saved_tensors_hooks`** （[more detail](https://github.com/meta-pytorch/torchtune/blob/10c31c0abadf51dfa2bf606637cd81e812e3f8c9/torchtune/training/_activation_offloading.py#L24-L374)）.
@stas00 @soulitzer

**Comment by stas00** [comment]
It was just a quick hack to demonstrate that it's a powerful technique - I just took the pytorch checkpoint function and bolted offloading onto it ;)

I'm not sure which of the re-entrant variants should be used. But again the point of the code I linked to wasn't to suggest any particular direction, I was just showing that w/o prefetch and streams it's a trivial change to the original core function. Perhaps if those could be abstracted away it'll be still quite simple if the logic remains the same and there will be some wrapper reading/writing the activation tensors and could optionally use non-hbm storage when assigning/retrieving (and prefetch). Which is sort of what tensor hooks implement.

I wonder if there can be made a new type of device which magically offloads/fetches tensor storage - with some triggers for an early prefetch. So once an activation tensor is moved to that device, it will automatically offload it, and when it's used it'd fetch it - with some sort of registered trigger to prefetch earlier than blocking real time. This is what FSDP and Deepspeed ZeRO perform already, including prefetch, except they shard the tensor across gpus. So this would be very similar except it'd shuttle the storage off to cpu or nvme or tape ;)

Additionally new gpus like gb200/300 have a very fast link of gpu <-> cpu memory (NVLink-C2C), so even the blocking version would be extremely fast there (if that link is used).

**Comment by ezyang** [comment]
cc @anijain2305

**Comment by alint77** [comment]
@stas00 @niyunsheng  @soulitzer  got a proof of concept working, tested on GH200, achieves 280GBps D2H and 350GBps H2D transfers, all overlapped with compute:

<img width="751" height="448" alt="Image" src="https://github.com/user-attachments/assets/26c4d019-d81e-46d6-a7e0-cfe00887b110" />

https://github.com/alint77/pytorch/tree/add/async-cpu-activation-offload

feedbacks are welcome

**Comment by alint77** [comment]
this is only applied to up_proj inside mlp btw since that's the largest activation...

**Comment by stas00** [comment]
That's awesome, @alint77 - it's probably the best to make a PR and discuss the implementation details there, no?

**Comment by alint77** [comment]
Well I'm new to pytorch iternals and still learning as I go, the POC was mostly vibecoded just to get a working prototype going, I will spend more time learning the internals and modifying the implementation to get it right (it's causing some recomputation in fwd pass atm), but the motivation is there because I'm working on Genome language model training optimization on the Jupiter cluster with GH200 nodes which would benefit massively from this because of ultra long contexts associated with genome models.

**Comment by alint77** [comment]
@stas00 PR is up, feedback is much appreciated. I'm getting great results on GH200, full transfer-compute overlap with mem usage being same as AC.

**Comment by alint77** [comment]
also have you seen this? https://poolside.ai/blog/tools-of-the-trade-c2c-activation-offloading-on-grace-blackwell
